### ISSUE TITLE
How to  support distributed training with huge embedding weights

### ISSUE THREAD

**Issue by adoda** [ray-project/ray#3554]
### System information
- no

### Describe the problem

We usually need to encode features in our models, such as RNN, NLP, ...
For example:

- The feature's embedding dict may be huge, it will be 300GB~1TB in our inhouse model.
- The weights can't be placed  into single node, so we will split the weights into parts, and place into each ps (parameter server).
- For computing loss, we used tf.nn.sampled_softmax_loss to take same samples and compute loss.
- The gradient is sparse for tf.nn.sampled_softmax_loss.

```python
softmax_w = tf.get_variable('softmax_w', [vsize, dec_size], dtype=tf.float32,
                       initializer=tf.truncated_normal_initializer(stddev=1e-4),
                       partitioner=tf.fixed_size_partitioner(num_shards=num_ps))
softmax_b = ...
...
loss = tf.nn.sampled_softmax_loss(
                            weights=softmax_w, biases=softmax_b, labels=labels, inputs=inputs,
                            num_sampled=num_samples, num_classes=vsize)
```

- So, how to implement this with ray? There is already distributed SGD in ray, but is still not obvious. 
- Is efficient when applying with sparse gradient? Need to copy the entire weights?

**Comment by ericl** [comment]
@robertnishihara this might be a good use case for the distributed tensorflow integration / differentiable Ray work you were looking at.

**Comment by adoda** [comment]
Any suggestion or update？ @robertnishihara @ericl

**Comment by simon-mo** [comment]
cc @richardliaw  for ray sgd

**Comment by richardliaw** [comment]
This is possible in Torch and possibly in TF (https://github.com/tensorflow/mesh). If there is a user request to enable this in RaySGD, we can open a new issue.

Thanks!
